### ISSUE TITLE
roslaunch --ros-args fails for required args

### ISSUE THREAD

**Issue by dornhege** [ros/ros_comm#231]
```
$ roscd roslaunch/test/xml/
$ roslaunch --ros-args test-arg.xml

arg 'required' is not defined. 

Arg xml is <arg name="param1_value" value="$(arg required)"/>
```

This is probably due to the usual roslaunch parsing code. In this case the error should be suppressed and still list all the args.

**Comment by dlaz** [comment]
This is happening because calling `roslaunch.config.load_config_default` tries to resolve the substitution args, but `required` doesn't have a value. A simple fix would be to add a keyword arg to `roslaunch.config.load_config_default` and pass it all the way down to `xmlloader.XmlLoader.resolve_args` so it's possible to tell resolve_args not to try to resolve substitution args. Unless there are any objections, I'll write a fix and submit a pull request.

Note: this bug also affects the --args parameter for roslaunch.

**Comment by dlaz** [comment]
Actually, this is a little more complicated. `arg_dump.py` relies on `loader.root_context.resolve_dict['arg']` containing all of the arguments in a launch file. The problem is that when adding args, anything with empty value or default is ignored. With just the fix in my last comment, the error will go away, but the arg `required` will not be displayed. I see two options:
- When the arg to resolve substitution args is False, give all some value that evaluates to True
- Maintain a separate arg list that doesn't care about values

I don't really like either of these options, so any suggestions would be appreciated.

**Comment by dirk-thomas** [comment]
I briefly looked into this and making it work for all cases will be pretty tough. I tried to use a mock dictionary to "return" arbitrary arguments but even that fails in numerous places. I will mark it as untargeted for now - patches are highly welcome.

**Comment by dlaz** [comment]
@dirk-thomas, is your mock dictionary code somewhere I could take a look at it? Do you have a test case that it fails on?

**Comment by dirk-thomas** [comment]
I ditched the code after the roslaunch-complete script failed to work on most of the launch files. Essentially where the LaunchContext is created I replaced the attributes `resolve_dict` and `include_resolve_dict` with a custom class which mimics a dict. It looked something like this:

```
class Mock(object):
    def __init__(self, attributes=None)
        self._attributes = attributes if attributes else {}
    def __contains__(self, key):
        if key not in self._attributes:
            self._attributes[key] = 'True'
        return True
    def __item__(self, key):
        self.__contains__(key)
        return self._attributes[key]
    def items(self):
        return self._attributes
```

**Comment by dirk-thomas** [comment]
Is this resolved with #230 being merged?

**Comment by jbohren** [comment]
> Is this resolved with #230 being merged?

No, but it is resolved by #379

**Comment by dirk-thomas** [comment]
Closing since it should be resolved quite some time ago.
